File: next_frontend/models/template.ts
// Define template config types
type ConfigType = "grid_based" | "cluster_based";

// Define template status
type TemplateStatus = "queued" | "processing" | "completed" | "failed";

interface Template {
  id: number;
  name: string;
  status: TemplateStatus;
  description?: string | null;
  config_type: ConfigType;
  configuration_file_id?: number | null;
  template_file_id?: number | null;
  num_questions: number;
  num_of_columns?: number | null;
  num_of_rows_per_column?: number[] | null;
  num_of_options_per_question: number;
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  created_by: number;
}

export type { Template, TemplateStatus, ConfigType };

File: fastapi_backend/app/schemas/template.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional

from app.models.template import TemplateConfigStatus, TemplateConfigType

class TemplateCreate(BaseModel):
    name: str
    description: Optional[str] = None
    config_type: TemplateConfigType
    template_file_id: int
    save_intermediate_results: bool = False
    
    # Additional fields for clustering-based configuration
    num_of_columns: Optional[int] = None
    num_of_rows_per_column: Optional[int] = None
    num_of_options_per_question: Optional[int] = None

class TemplateResponse(BaseModel):
    id: int
    name: str
    status: TemplateConfigStatus
    description: Optional[str] = None
    config_type: TemplateConfigType
    configuration_file_id: Optional[int] = None
    template_file_id: Optional[int] = None
    num_questions: int
    options_per_question: int
    created_at: datetime
    updated_at: datetime
    created_by: int